nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+        dma_start_fence();
+        // enter TX mode
+        self.radio.tasks_txen.write(|w| w.tasks_txen().set_bit());
+
+        // due to a shortcut the transmission will start automatically so we just have to wait
+        // until the PHYEND event is raised
+        self.wait_for_event(Event::PhyEnd);
+        dma_end_fence();
+
+        Ok(())
+    }
+
+    /// Sends the given `packet`
+    ///
+    /// This is utility method that *consecutively* calls the `try_send` method until it succeeds.
+    /// Note that this approach is *not* IEEE spec compliant -- there must be delay between failed
+    /// CCA attempts to be spec compliant
+    pub fn send(&mut self, packet: &Packet) {
+        // NOTE we do NOT check the address of `packet`; see comment in `Packet::new` for details
+
+        self.radio.shorts.modify(|_, w| w.ccaidle_txen().set_bit());
+
+        // go to the RXIDLE state
+        self.enable_rx();
+
+        // NOTE the DMA doesn't exactly start at this point but due to shortcuts it may occur at any
+        // point after this volatile write
+        dma_start_fence();
+        // NOTE(unsafe) DMA transfer has not yet started
+        unsafe {
+            self.radio
+                .packetptr
+                .write(|w| w.packetptr().bits(packet.buffer.as_ptr() as u32));
+        }
+
+        'cca: loop {
+            // start CCA
+            self.radio
+                .tasks_ccastart
+                .write(|w| w.tasks_ccastart().set_bit());
+
+            loop {
+                if self
+                    .radio
+                    .events_ccaidle
+                    .read()
+                    .events_ccaidle()
+                    .bit_is_set()
+                {
+                    // channel is clear
+                    self.radio.events_ccaidle.reset();
+                    break 'cca;
+                }
+
+                if self
+                    .radio
+                    .events_ccabusy
+                    .read()
+                    .events_ccabusy()
+                    .bit_is_set()
+                {
+                    // channel is busy
+                    self.radio.events_ccaidle.reset();
+                    continue 'cca;
+                }
+            }
+        }
+
+        // due to a shortcut the transmission will start automatically so we just have to wait
+        // until the PHYEND event is raised
+        self.wait_for_event(Event::PhyEnd);
+        dma_end_fence();
+
+        self.radio
+            .shorts
+            .modify(|_, w| w.ccaidle_txen().clear_bit());
+    }
+
+    /// Moves the radio from any state to the DISABLED state
+    fn disable(&mut self) {
+        // See figure 110 in nRF52840-PS
+        loop {
+            if let Variant::Val(state) = self.radio.state.read().state().variant() {
+                match state {
+                    STATE_A::DISABLED => return,
+
+                    STATE_A::RXRU | STATE_A::RXIDLE | STATE_A::TXRU | STATE_A::TXIDLE => {
```

Comment:
`RXIDLE` state may indicate ongoing CCA procedure that may result in TX with radio shorts. To be fully safe, following set of tasks should be triggered:
`CCASTOP`
`STOP`
`DISABLE`